tools/windows-node-installer/test/e2e/aws_test.go

```diff
@@ -1,90 +1,329 @@
+func updateCreatedInstance(instanceID string) (*ec2.Instance, error) {
+	instances, err := awsProvider.EC2.DescribeInstances(&ec2.DescribeInstancesInput{
+		InstanceIds: aws.StringSlice([]string{instanceID}),
+	})
+
 	if err != nil {
-		t.Fatalf("error while getting infrastructure ID for the given OpenShift cluster")
+		return nil, err
 	}
-	if !checkIfInfraTagExists(instance, infraID) {
-		t.Fatalf("Infrastructure tag not available for %v", info.InstanceIDs[0])
+	if len(instances.Reservations) < 1 || len(instances.Reservations[0].Instances) < 1 {
+		return nil, fmt.Errorf("instance does not exist")
 	}
-	err = cloud.DestroyWindowsVMs()
-	assert.NoError(t, err, "error deleting instance")
+	return instances.Reservations[0].Instances[0], err
+}
 
-	// the windows-node-installer.json should be removed after resource is deleted.
-	_, err = resource.ReadInstallerInfo(dir)
-	assert.Error(t, err, "error deleting windows-node-installer.json file")
+// tearDownInstance removes the lingering resources including instance and Windows security group when required steps of
+// the test fail.
+func tearDownInstance(t *testing.T, Msg string, terr error) {
+	t.Logf("%s, tearing down lingering resources", Msg)
+
+	if createdInstanceID != "" {
+		err := awsProvider.TerminateInstance(createdInstanceID)
+		if err != nil {
+			t.Errorf("error terminating instance during teardown, %v", err)
+		}
+	}
+
+	if createdSgID != "" {
+		err := awsProvider.DeleteSG(createdSgID)
+		if err != nil {
+			t.Errorf("error terminating instance during teardown, %v", err)
+		}
+	}
+	assert.FailNow(t, terr.Error(), Msg)
+}
+
+// checkInstanceProperties updates the createdInstance global object and asserts if an instance is in the running
+// state, has the right image id, instance type, and ssh key associated.
+func checkInstanceProperties(t *testing.T) {
+	instance, err := updateCreatedInstance(createdInstanceID)
+	if err != nil {
+		tearDownInstance(t, "error getting ec2 instance object after creation", err)
+	} else {
+		createdInstance = instance
+	}
+	assert.Equal(t, ec2.InstanceStateNameRunning, *createdInstance.State.Name, "instance is not in running state")
+
+	assert.Equalf(t, imageID, *createdInstance.ImageId, "instance image ID: %s", *createdInstance.ImageId)
+
+	assert.Equalf(t, instanceType, *createdInstance.InstanceType, "instance type %s", *createdInstance.InstanceType)
+
+	assert.Equalf(t, sshKey, *createdInstance.KeyName, "instance ssh key name: %s", *createdInstance.KeyName)
+}
+
+// checkInstanceHasPublicSubnetAndIp asserts if the instance is associated with a public IP address and  subnet by
+// checking if the subnet routing table has internet gateway attached.
+func checkInstanceHasPublicSubnetAndIp(t *testing.T) {
+	assert.NotEmpty(t, createdInstance.PublicIpAddress, "instance does not have a public IP address")
+
+	routeTables, err := awsProvider.EC2.DescribeRouteTables(&ec2.DescribeRouteTablesInput{
```

Comment:
I'd recommend testing our code here, for example we can get the subnet, compare it with subnet that we get from aws package, same goes with public IP. check if the public we're getting associated with the node is same as what we expect it to be.

Reply:
We do not have a source of truth for subnet ID or IP. The IP is auto-assigned therefore if it is not empty we are good. For subnet, we are randomly grabbing a public subnet and therefore, we check if the subnet that is attached to our instance has igw in the route table to determine if it is public.